tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs

```diff
@@ -1672,6 +1672,65 @@ public async Task ValidateGetAccountAsyncWithNullEmptyAccountIdAsync(string acco
                 Assert.IsNull(acc);
             }
         }
+
+        [TestMethod]
+        public async Task ValidateAppTokenProviderAsync()
+        {
+            var app = ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                          .WithAppTokenProvider(async (AppTokenProviderParameters parameters) =>
+                                                          {
+                                                              Assert.IsNotNull(parameters.Scopes);
+                                                              Assert.IsNotNull(parameters.CorrelationId);
+                                                              Assert.IsNotNull(parameters.TenantId);
+                                                              Assert.IsNotNull(parameters.CancellationToken);
+
+                                                              return Task.Run(() => GetAppTokenProviderResult()).Result;
+                                                          })
+                                                          .BuildConcrete();
+
+            //AcquireToken from app provider
```

Comment:
Wonder if we should add a new token source maybe? ExternalTokenProvider? Or at least maybe log where the token came from. I think currently we log the token endpoint, but in this case it would be null, so from the logs if would seem like this is a cached token.

Reply:
Not sure about adding a new token source, because it'll confuse 99% of users who don't use this status. `IdentityProvider` seems ok in this case. Agreed that we need some logging statements to make sure this is set.